raft/node_test.go

```diff
@@ -142,6 +142,54 @@ func TestNodePropose(t *testing.T) {
 	}
 }
 
+// TestNodeReadIndex ensures that node.ReadIndex sends the MsgReadIndex message to the underlying raft.
+// it also ensures the ReadIndexState can be out through ready channel
+func TestNodeReadIndex(t *testing.T) {
+	msgs := []raftpb.Message{}
+	appendStep := func(r *raft, m raftpb.Message) {
+		msgs = append(msgs, m)
+	}
+
+	n := newNode()
+	s := NewMemoryStorage()
+	r := newTestRaft(1, []uint64{1}, 10, 1, s)
+	r.readIndex = 1
+	r.readIndexCtx = []byte("somedata")
+	go n.run(r)
+	n.Campaign(context.TODO())
+	for {
+		rd := <-n.Ready()
+		if rd.ReadIndex != 1 {
+			t.Errorf("ReadIndex = %d, want 1", rd.ReadIndex)
+		}
+
+		if !reflect.DeepEqual(rd.ReadIndexCtx, []byte("somedata")) {
+			t.Errorf("ReadIndexCtx = %v, want %v", rd.ReadIndexCtx, []byte("somedata"))
+		}
+
+		s.Append(rd.Entries)
+		// change the step function to appendStep until this raft becomes leader
+		if rd.SoftState.Lead == r.id {
+			r.step = appendStep
```

Comment:
let's break out first and change r.step later around line 179. so it would be closer to the checking. and code would be more readable.